Add DLQ and aggregate processor edits. Baseline processor names (#3880)
opensearch-project/documentation-website: _data-prepper/pipelines/dlq.md

---
layout: default
title: Dead-letter queues 
parent: Pipelines
nav_order: 13
---

# Dead-letter queues

> Data Prepper pipelines support dead-letter queues (DLQ) to offload failed events and make them accessible for analysis.

Reviewer: Should this be "DLQs"?

As of Data Prepper 2.3, one the `s3` source supports DLQs.

## Configure a DLQ writer

> To configure a DLQ writer for the `s3` source, add the following to your pipeline.yaml's `sink` configuration:

Reviewer: Pretty sure this is `pipelines.yaml`

```yaml
  sink:
    opensearch:
      dlq:
        s3:
          bucket: "my-dlq-bucket"
          key_path_prefix: "dlq-files/"
          region: "us-west-2"
          sts_role_arn: "arn:aws:iam::123456789012:role/dlq-role"
```

The resulting DLQ file outs as a JSON array of DLQ objects. Any files written to the S3 DLQ contains the following name pattern:

```
dlq-v${version}-${pipelineName}-${pluginId}-${timestampIso8601}-${uniqueId}
```

- `version`: The Data Prepper version.
- `pipelineName`: The pipeline name indicated in pipeline.yaml.
- `pluginId`: The ID of the plugin associated with the DLQ event.

## Configuration

DLQ supports the following configuration options.

Option | Required | Type | Description
:--- | :--- | :--- | :---

> bucket | Yes | String | The bucket name of the DLQ outputs failed records.

Reviewer: output's?

Author: Switched to The bucket name where the DLQ outputs failed records.

region | No | String | The AWS Region of the S3 Bucket. Defaults to `us-east-1`.
sts_role_arn | No | The STS role the DLQ assumes in order to write to an AWS S3 bucket. Default is `null`, which uses the standard SDK behavior for credentials. To use this option, the S3 bucket must have the `S3:PutObject` permission configured.

## Metrics

DLQ supports the following metrics

### Counter

- `dlqS3RecordsSuccess`: Measures the number of successful records sent to S3.
- `dlqS3RecordsFailed`: Measures the number of failed records failed to be sent to S3.
- `dlqS3RequestSuccess`: Measures the number of successful S3 requests.
- `dlqS3RequestFailed`: Measures the number of failed S3 requests.

### Distribution summary

- `dlqS3RequestSizeBytes`: Measures the distribution of the S3 request's payload size in bytes.

### Timer

- `dlqS3RequestLatency`: Measures latency when sending each S3 request including retries.

## DLQ Objects

DQL supports the following DLQ objects:

> * `pluginId` : The ID of the plugin which resulted in the event being DLQ’ed.

Reviewer: Maybe "in the event being sent to the DLQ"?

Author: "The ID of the plugin where the event sent to the DLQ originated."

---
layout: default
title: Dead-letter queues 
parent: Pipelines
nav_order: 13
---

# Dead-letter queues

> Data Prepper pipelines support dead-letter queues (DLQs) to offload failed events and make them accessible for analysis.

Reviewer 2: ```suggestion
Data Prepper pipelines support dead-letter queues (DLQs) for offloading failed events and making them accessible for analysis.
```

> As of Data Prepper 2.3, only the `s3` source supports DLQs.

Reviewer 2: 2.2?

Author: 2.3, its a new feature.

## Configure a DLQ writer

To configure a DLQ writer for the `s3` source, add the following to your pipeline.yaml:

```yaml
  sink:
    opensearch:
      dlq:
        s3:
          bucket: "my-dlq-bucket"
          key_path_prefix: "dlq-files/"
          region: "us-west-2"
          sts_role_arn: "arn:aws:iam::123456789012:role/dlq-role"
```

> The resulting DLQ file outs as a JSON array of DLQ objects. Any files written to the S3 DLQ contains the following name pattern:

Reviewer 3: ```suggestion
The resulting DLQ file outputs as a JSON array of DLQ objects. Any file written to the S3 DLQ contains the following name pattern:
```

Reviewer 2: Agree

```
dlq-v${version}-${pipelineName}-${pluginId}-${timestampIso8601}-${uniqueId}
```

> 

Author: ```suggestion
The following information is replaced in the name pattern:

```

> - `version`: The Data Prepper version.

Reviewer 2: The list should be introduced by a brief sentence ending in a colon.

- `pipelineName`: The pipeline name indicated in pipeline.yaml.
- `pluginId`: The ID of the plugin associated with the DLQ event.

## Configuration

DLQ supports the following configuration options.

Option | Required | Type | Description
:--- | :--- | :--- | :---
bucket | Yes | String | The bucket name where the DLQ outputs failed records.

> key_path_prefix | No | String | The `key_prefix` to use in the S3 bucket. Defaults to `""`. Supports time value pattern variables such as `/%{yyyy}/%{MM}/%{dd}`, including any variables listed in the [Java DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html). For example, when using the `/%{yyyy}/%{MM}/%{dd}` pattern, you can set `key_prefix` as `/2023/01/24`.

Reviewer 2: ```suggestion
key_path_prefix | No | String | The `key_prefix` used in the S3 bucket. Defaults to `""`. Supports time value pattern variables, such as `/%{yyyy}/%{MM}/%{dd}`, including any variables listed in the [Java DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html). For example, when using the `/%{yyyy}/%{MM}/%{dd}` pattern, you can set `key_prefix` as `/2023/01/24`.
```

> region | No | String | The AWS Region of the S3 Bucket. Defaults to `us-east-1`.

Reviewer 2: ```suggestion
region | No | String | The AWS Region of the S3 bucket. Defaults to `us-east-1`.
```

> sts_role_arn | No | The STS role the DLQ assumes in order to write to an AWS S3 bucket. Default is `null`, which uses the standard SDK behavior for credentials. To use this option, the S3 bucket must have the `S3:PutObject` permission configured.

Reviewer 2: ```suggestion
sts_role_arn | No | The AWS STS role the DLQ assumes in order to write to an S3 bucket. Default is `null`, which uses the standard SDK behavior for credentials. To use this option, the S3 bucket must have the `S3:PutObject` permission configured.
```

> sts_role_arn | No | The STS role the DLQ assumes in order to write to an AWS S3 bucket. Default is `null`, which uses the standard SDK behavior for credentials. To use this option, the S3 bucket must have the `S3:PutObject` permission configured.

Reviewer 4: Check the columns - I think the description is in the wrong place.

> When using DLQ with an opensearch sink, you can configure the [max_retries]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#configure-max_retries) option to send failed data to the DLQ when the sink reaches the maximum number of retries.

Reviewer 2: ```suggestion
When using DLQ with an OpenSearch sink, you can configure the [max_retries]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#configure-max_retries) option to send failed data to the DLQ when the sink reaches the maximum number of retries.
```

## Metrics

> DLQ supports the following metrics

Reviewer 4: Add a colon after metrics

> DLQ supports the following metrics

Reviewer 2: ```suggestion
DLQ supports the following metrics.
```

> When using DLQ with an [opensearch]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/) sink, you can configure the [max_retries]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#configure-max_retries) option to send failed data to the DLQ when the sink reaches the maximum number of retries.

Reviewer 3: ```suggestion
When using DLQ with an [opensearch sink]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/), you can [configure the max_retries]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/opensearch/#configure-max_retries) option to send failed data to the DLQ when the sink reaches the maximum number of retries.
```

### Counter

- `dlqS3RecordsSuccess`: Measures the number of successful records sent to S3.

> - `dlqS3RecordsFailed`: Measures the number of failed records failed to be sent to S3.

Reviewer 3: ```suggestion
- `dlqS3RecordsFailed`: Measures the number of records that failed to be sent to S3.
```

Reviewer 2: Agree

- `dlqS3RequestSuccess`: Measures the number of successful S3 requests.
- `dlqS3RequestFailed`: Measures the number of failed S3 requests.

### Distribution summary

- `dlqS3RequestSizeBytes`: Measures the distribution of the S3 request's payload size in bytes.

### Timer

> - `dlqS3RequestLatency`: Measures latency when sending each S3 request including retries.

Reviewer 2: ```suggestion
- `dlqS3RequestLatency`: Measures latency when sending each S3 request, including retries.
```

> ## DLQ Objects

Reviewer 2: ```suggestion
## DLQ objects
```

> DQL supports the following DLQ objects:

Reviewer 2: DLQ?

> DQL supports the following DLQ objects:

Author: ```suggestion
DLQ supports the following DLQ objects:
```

> * `pluginId` : The ID of the plugin where the event sent to the DLQ originated.

Reviewer 2: ```suggestion
* `pluginId`: The ID of the plugin that originated the event sent to the DLQ.
```

> * `pluginName` : The name of the plugin.

Reviewer 2: ```suggestion
* `pluginName`: The name of the plugin.
```

* `failedData` : An object that contains the failed object and it's options. This object is unique to each plugin.

> * `pipelineName` : The name of the Data Prepper pipeline where the event failed.

Reviewer 2: ```suggestion
* `pipelineName`: The name of the Data Prepper pipeline where the event failed.
```

> * `pipelineName` : The name of the Data Prepper pipeline where the event failed.

Reviewer 2: "in which" instead of "where"?

> * `failedData` : An object that contains the failed object and it's options. This object is unique to each plugin.

Reviewer 3: ```suggestion
* `failedData` : An object that contains the failed object and its options. This object is unique to each plugin.
```

Reviewer 2: Removed a space.

> * `timestamp` : The timestamp of the failures in an `ISO8601` format.

Reviewer 2: ```suggestion
* `timestamp`: The timestamp of the failures in an `ISO8601` format.
```
